deps/core.lua

```diff
@@ -238,45 +229,56 @@ function Emitter:emit(name, ...)
     self:missingHandlerType(name, ...)
     return
   end
-  local handlers_for_type = rawget(handlers, name)
-  if not handlers_for_type then
+
+  handlers = rawget(handlers, name)
+  if handlers then
+    -- avoid be removed
+    local _ = {}
+    for i=1, #handlers do
+      _[i] = handlers[i]
+    end
+    for i=1, #_ do
+      _[i](...)
+    end
+  else
     self:missingHandlerType(name, ...)
     return
   end
-  for i = 1, #handlers_for_type do
-    local handler = handlers_for_type[i]
-    if handler then handler(...) end
-  end
-  for i = #handlers_for_type, 1, -1 do
-    if not handlers_for_type[i] then
-      table.remove(handlers_for_type, i)
-    end
-  end
+
   return self
 end
 
 -- Remove a listener so that it no longer catches events.
 function Emitter:removeListener(name, callback)
   local handlers = rawget(self, "handlers")
   if not handlers then return end
-  local handlers_for_type = rawget(handlers, name)
-  if not handlers_for_type then return end
+
+  if not name then
+    rawset(self, "handlers", {})
+    return
+  end
+
   if callback then
-    for i = #handlers_for_type, 1, -1 do
-      local h = handlers_for_type[i]
-      if type(h) == "function" then
-        h = h == callback
-      elseif type(h) == "table" then
-        h = h == callback or h.callback == callback
+    handlers = rawget(handlers, name)
+    if not handlers then return end
+
+    for i = #handlers, 1, -1 do
+      local h = handlers[i]
+      if type(callback) == "table" then
+        -- only remove once callback table
+        h = type(h) == 'table' and h == callback or false
+      elseif type(callback) == "function" then
+        -- remove once callback table or other callback function
+        h = type(h) == 'table' and h.callback == callback or h == callback
       end
       if h then
-        handlers_for_type[i] = false
```

Comment:
Emitter:removeListener `handlers_for_type[i] = false` only affect the libuv next cycle.
Direct remove it is safe?

Reply:
Should I not change this?